src/dataflow/src/render/mod.rs

```diff
@@ -255,17 +275,70 @@ pub fn build_dataflow<A: Allocate>(
                         source_metrics,
                     );
 
+                let ok_handle = Rc::new(EventLink::new());
+                let err_handle = Rc::new(EventLink::new());
+
+                results.insert(
+                    *src_id,
+                    SourceBoundary {
+                        ok: Rc::<_>::clone(&ok_handle),
+                        err: Rc::<_>::clone(&err_handle),
+                        token: source_token,
+                        additional_tokens,
+                    },
+                );
+
+                ok.inner.capture_into(ok_handle);
+                err.inner.capture_into(err_handle);
+            }
+        })
+    });
+
+    results
+}
+
+/// Assemble the "compute"  side of a dataflow, i.e. all but the sources.
+///
+/// This method imports sources from provided assets, and then builds the remaining
+/// dataflow using "compute-local" assets like shared arrangements, and producing
+/// both arrangements and sinks.
+pub fn build_compute_dataflow<A: Allocate>(
+    timely_worker: &mut TimelyWorker<A>,
+    compute_state: &mut ComputeState,
+    sources: BTreeMap<GlobalId, SourceBoundary>,
+    dataflow: DataflowDescription<plan::Plan>,
+    sink_metrics: &SinkBaseMetrics,
+) {
+    let worker_logging = timely_worker.log_register().get("timely");
+    let name = format!("Dataflow: {}", &dataflow.debug_name);
+
+    timely_worker.dataflow_core(&name, worker_logging, Box::new(()), |_, scope| {
+        // The scope.clone() occurs to allow import in the region.
+        // We build a region here to establish a pattern of a scope inside the dataflow,
+        // so that other similar uses (e.g. with iterative scopes) do not require weird
+        // alternate type signatures.
+        scope.clone().region_named(&name, |region| {
+            let mut context = Context::for_dataflow(&dataflow, scope.addr().into_element());
+            let mut tokens = RelevantTokens::default();
+
+            // Import declared sources into the rendering context.
+            for (source_id, source) in sources.into_iter() {
                 // Associate collection bundle with the source identifier.
-                context.insert_id(Id::Global(*src_id), collection_bundle);
+                let ok = Some(source.ok).replay_into(region).as_collection();
+                let err = Some(source.err).replay_into(region).as_collection();
```

Comment:
This has the undesirable side-effect that we busy-wait on new data :/ We certainly need to pass activations around.